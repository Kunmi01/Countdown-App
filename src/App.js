import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import {
  supportsLocalStorage,
  savePromoData,
  loadPromoData,
  savePromoEndDate,
  loadPromoEndDate
} from './utils/localStorage';

import Countdown from './components/Countdown';
import topImage from './assets/images/top-image.png';
import topImageRetina from './assets/images/top-image-2x.png';
import './App.scss';

const COUNTDOWN_REFRESH_INTERVAL = 500;

const App = ({ apiUrl }) => {
  const [promoData, setPromoData] = useState();
  const [timeRemaining, setTimeRemaining] = useState(0);
  const [countdownInterval, setCountdownInterval] = useState();
  const [errorMessages, setErrorMessages] = useState([]);

  const isDataValid = ({ duration, cashValue, optInUrl }) => {
    const errors = [];
    if (!duration || typeof duration !== 'number') {
      errors.push('Missing or invalid duration value');
    }
    if (!cashValue) {
      errors.push('Missing cash value');
    }
    if (!optInUrl) {
      errors.push('Missing opt in url');
    }

    setErrorMessages([...errorMessages, ...errors]);
    return !errors.length;
  };

  const hasLocalStorage = () => {
    if (supportsLocalStorage()) {
      return true;
    }
    setErrorMessages([
      ...errorMessages,
      'Browser does not support local storage'
    ]);
    return false;
  };

  const initializeCountdown = async () => {
    const storedData = loadPromoData();
    const storedEndDate = loadPromoEndDate();
    let jsonData;

    try {
      jsonData = await (await fetch(apiUrl)).json();
    } catch (err) {
      setErrorMessages([...errorMessages, err.message]);
      return;
    }

    if (storedData && storedData.id === jsonData.id && storedEndDate) {
      const getMsLeft = () => storedEndDate - Date.now();
      setPromoData(storedData);
      setTimeRemaining(getMsLeft());
      setCountdownInterval(
        setInterval(() => {
          setTimeRemaining(getMsLeft());
        }, COUNTDOWN_REFRESH_INTERVAL)
      );
    } else if (isDataValid(jsonData) && hasLocalStorage()) {
      const endDate = Date.now() + jsonData.duration;
      savePromoData(jsonData);
      savePromoEndDate(endDate);
      initializeCountdown();
    }
  };

  useEffect(() => {
    initializeCountdown();
  }, []);

  useEffect(() => {
    if (timeRemaining <= 1000) {
      clearInterval(countdownInterval);
      setTimeRemaining(0);
    }
    // console.log(timeRemaining);
  }, [timeRemaining]);

  return (
    <div className="app">
      {!!errorMessages.length &&
        errorMessages.map(message => (
          <div className="app__error-message" key={message}>
            {message}
          </div>
        ))}
      <div className="app__wrapper">
        {!!timeRemaining && (
          <>
            <div className="app__image-wrapper">
              <picture>
                <source srcSet={`${topImage} 1x, ${topImageRetina} 2x`} />
                <img src={topImage} alt="top" />
              </picture>
            </div>
            <h3 className="app__message">
              Get your free <b>Â£{promoData?.cashValue}</b> now
            </h3>
          </>
        )}
        <div className="app__countdown-wrapper">
          <Countdown duration={timeRemaining} />
          {!!timeRemaining && (
            <a
              className="app__opt-in-button"
              href={promoData?.optInUrl}
              target="_blank"
              rel="noreferrer"
            >
              Opt in
            </a>
          )}
        </div>
      </div>
    </div>
  );
};

App.propTypes = {
  apiUrl: PropTypes.string.isRequired
};

export default App;
