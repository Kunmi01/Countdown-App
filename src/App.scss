@import './scss/base.scss';

$TOP_IMAGE_MAX_WIDTH: 500px;
$COUNTDOWN_MAX_WIDTH: 750px;

.app {
  display: flex;
  flex-direction: column;
  justify-content: center;
  color: $white;
  min-height: 100vh;

  &__error-message {
    background: rgba($color: $black, $alpha: 0.8);
    color: $red;
    padding: $unit * 2;
  }

  &__wrapper {
    @include component-wrapper;
    text-align: center;
  }

  &__image-wrapper {
    max-width: $TOP_IMAGE_MAX_WIDTH;
    margin: 0 auto $unit * 8;

    img {
      display: block;
      width: 100%;
    }
  }

  &__message {
    margin-bottom: $unit * 5;

    @include bp-from('mobile') {
      font-size: $font-xxl;
    }
  }

  &__countdown-wrapper {
    max-width: $COUNTDOWN_MAX_WIDTH;
    margin: 0 auto;
  }

  &__opt-in-button {
    display: block;
    border: 2px solid $neon-green;
    border-radius: $unit;
    box-shadow: inset 0 0 10px $deep-green, 0 0 10px $deep-green;
    background: rgba($color: $black, $alpha: 0.5);
    font-family: $font-play-bold;
    font-size: $font-l;
    color: $white;
    text-decoration: none;
    width: 100%;
    opacity: 0.88;
    padding: $unit * 4;
    margin-top: $unit * 10;
    cursor: pointer;

    &:hover {
      opacity: 1;
    }
  }
}
