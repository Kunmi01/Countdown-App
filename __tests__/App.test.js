import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import App from '../src/App';

describe('App', () => {
  const wrapper = shallow(<App apiUrl="http://localhost:4000/promo" />);

  it('renders App component', () => {
    expect(wrapper).toHaveLength(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
