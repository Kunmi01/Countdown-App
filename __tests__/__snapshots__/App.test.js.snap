// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App component 1`] = `
<div
  className="app"
>
  <div
    className="app__wrapper"
  >
    <div
      className="app__countdown-wrapper"
    >
      <Countdown
        duration={0}
      />
    </div>
  </div>
</div>
`;
